### ISSUE TITLE
'Attempt to free temp prematurely' from Devel::Peek::DumpProg()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12881]
Migrated from [rt.perl.org#117353](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117353) (status was 'rejected')

Searchable as RT117353$

**Comment by p5pRT** [comment]
### From @Hugmeir

#### Created by @Hugmeir



Looks like two uses of SvREFCNT\_dec\_NN\(\) survived
2b656fcc48f28912136698c28b3bd916c42d74f8\, which are causing
DumpProg\(\)/op\_dump to show a bunch of 'Attempt to free temp
prematurely' warnings\. We don't have tests for DumpProg\(\) so this went
by unnoticed \-\- by luck I was just starting to test that and this came
up\. The attached patch resolves this but doesn't add tests\, leaving
us in the same situation as before\, although MAD is probably still broken\.
For a more throughout fix\, see
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117273

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.16.2:

Configured by hugmeir at Tue Nov 20 17:20:00 ART 2012.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:

  Platform:
    osname=linux, osvers=3.5.0-18-generic, archname=x86_64-linux-thread-multi
    uname='linux naw 3.5.0-18-generic #29-ubuntu smp fri oct 19
10:26:51 utc 2012 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de
-Dprefix=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2 -DDEBUGGING
-Dusethreads -Doptimize=-g -O0 -ggdb3 -Uversiononly -Accflags=-Wall
-Wextra -Aeval:scriptdir=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra
-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O0 -ggdb3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -O0 -ggdb3
-L/usr/local/lib -fstack-protector'

Locally applied patches:



@INC for perl 5.16.2:
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5/x86_64-linux-gnu-thread-multi
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5/x86_64-linux-thread-multi
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2
    .


Environment for perl 5.16.2:
    HOME=/home/hugmeir
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hugmeir/.rbenv/shims:/home/hugmeir/.rbenv/bin:/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin:/home/hugmeir/.rbenv/shims:/home/hugmeir/.rbenv/bin:/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL5LIB=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5/x86_64-linux-gnu-thread-multi:/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5
    PERLBREW_BASHRC_VERSION=0.46
    PERLBREW_HOME=/home/hugmeir/.perlbrew
    PERLBREW_LIB=all
    PERLBREW_MANPATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/man:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/man
    PERLBREW_PATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin
    PERLBREW_PERL=perl-5.16.2
    PERLBREW_ROOT=/home/hugmeir/perl5/perlbrew
    PERLBREW_VERSION=0.46
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MB_OPT=--install_base /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MM_OPT=INSTALL_BASE=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

<details><summary>0001-Removed-two-leftover-SvREFCNT_dec_NN-s-from-dump.c.patch</summary>

```diff
From 61dd5037b5974da2ad492f398d620dbaa97017b9 Mon Sep 17 00:00:00 2001
From: Brian Fraser <fraserbn@gmail.com>
Date: Sun, 24 Mar 2013 02:30:18 -0300
Subject: [PATCH] Removed two leftover SvREFCNT_dec_NN()s from dump.c

---
 dump.c |    2 --
 1 file changed, 2 deletions(-)

diff --git a/dump.c b/dump.c
index eab747c..1952ffc 100644
--- a/dump.c
+++ b/dump.c
@@ -864,7 +864,6 @@ S_op_private_to_names(pTHX_ SV *tmpsv, U32 optype, U32 op_private) {
         else                                                            \
             PerlIO_printf(file, " flags=\"%s\"",                        \
                           SvCUR(tmpsv) ? SvPVX(tmpsv) + 1 : "");        \
-        SvREFCNT_dec_NN(tmpsv);                                            \
     }
 
 #if !defined(PERL_MAD)
@@ -958,7 +957,6 @@ S_op_private_to_names(pTHX_ SV *tmpsv, U32 optype, U32 op_private) {
 	} else if (!xml)                                                \
             Perl_dump_indent(aTHX_ level, file, "PRIVATE = (0x%"UVxf")\n", \
                              (UV)oppriv);                               \
-	SvREFCNT_dec_NN(tmpsv);                                         \
     }
 
 
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Mar 25\, 2013 at 05&#8203;:29&#8203;:27PM \-0700\, Brian Fraser wrote&#8203;:
> Looks like two uses of SvREFCNT\_dec\_NN\(\) survived
> 2b656fcc48f28912136698c28b3bd916c42d74f8

I don't understand what you mean here\. That commit was a mistake&#8203;:
as well as my own fix\, it inadvertently included Reini's patch
from \#117273\. The whole commit was reverted with 17605be794\,
then my own small fix was re\-added with 09af213235\.

\-\- 
But Pity stayed his hand\. "It's a pity I've run out of bullets"\,
he thought\. \-\- "Bored of the Rings"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Hugmeir

On Tue\, Mar 26\, 2013 at 7&#8203;:44 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Mon\, Mar 25\, 2013 at 05&#8203;:29&#8203;:27PM \-0700\, Brian Fraser wrote&#8203;:
>> Looks like two uses of SvREFCNT\_dec\_NN\(\) survived
>> 2b656fcc48f28912136698c28b3bd916c42d74f8
>
> I don't understand what you mean here\. That commit was a mistake&#8203;:
> as well as my own fix\, it inadvertently included Reini's patch
> from \#117273\. The whole commit was reverted with 17605be794\,
> then my own small fix was re\-added with 09af213235\.
>

I actually sent this on Saturday\, but it got caught up in perlbug's
spam filter\, then didn't check whenever it had already been fixed
before i resent it\. Sorry for the noise\!

> \-\-
> But Pity stayed his hand\. "It's a pity I've run out of bullets"\,
> he thought\. \-\- "Bored of the Rings"

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'
